src/Evolution/Systems/NewtonianEuler/BoundaryCorrections/Hll.cpp

```diff
@@ -0,0 +1,294 @@
+              normal_covector);
+  for (size_t i = 0; i < Dim; ++i) {
+    packaged_normal_dot_flux_momentum_density->get(i) =
+        get<0>(normal_covector) * flux_momentum_density.get(0, i);
+    for (size_t j = 1; j < Dim; ++j) {
+      packaged_normal_dot_flux_momentum_density->get(i) +=
+          normal_covector.get(j) * flux_momentum_density.get(j, i);
+    }
+  }
+  dot_product(packaged_normal_dot_flux_energy_density, flux_energy_density,
+              normal_covector);
+
+  return fmax(max(get(*packaged_largest_outgoing_char_speed)),
+              -min(get(*packaged_largest_ingoing_char_speed)));
+}
+
+template <size_t Dim>
+void Hll<Dim>::dg_boundary_terms(
+    const gsl::not_null<Scalar<DataVector>*> boundary_correction_mass_density,
+    const gsl::not_null<tnsr::I<DataVector, Dim, Frame::Inertial>*>
+        boundary_correction_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> boundary_correction_energy_density,
+    const Scalar<DataVector>& mass_density_int,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& momentum_density_int,
+    const Scalar<DataVector>& energy_density_int,
+    const Scalar<DataVector>& normal_dot_flux_mass_density_int,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_momentum_density_int,
+    const Scalar<DataVector>& normal_dot_flux_energy_density_int,
+    const Scalar<DataVector>& largest_outgoing_char_speed_int,
+    const Scalar<DataVector>& largest_ingoing_char_speed_int,
+    const Scalar<DataVector>& mass_density_ext,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& momentum_density_ext,
+    const Scalar<DataVector>& energy_density_ext,
+    const Scalar<DataVector>& normal_dot_flux_mass_density_ext,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_momentum_density_ext,
+    const Scalar<DataVector>& normal_dot_flux_energy_density_ext,
+    const Scalar<DataVector>& largest_outgoing_char_speed_ext,
+    const Scalar<DataVector>& largest_ingoing_char_speed_ext,
+    const dg::Formulation dg_formulation) const noexcept {
+  // Allocate a temp buffer with four tags.
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>,
+                       ::Tags::TempScalar<2>, ::Tags::TempScalar<3>>>
+      temps{get(mass_density_int).size()};
+
+  // Determine lambda_max and lambda_min from the characteristic speeds info
+  // from interior and exterior
+  get(get<::Tags::TempScalar<0>>(temps)) =
+      max(0., get(largest_outgoing_char_speed_int),
+          -get(largest_ingoing_char_speed_ext));
+  const DataVector& lambda_max = get(get<::Tags::TempScalar<0>>(temps));
+  get(get<::Tags::TempScalar<1>>(temps)) =
+      min(0., get(largest_ingoing_char_speed_int),
+          -get(largest_outgoing_char_speed_ext));
+  const DataVector& lambda_min = get(get<::Tags::TempScalar<1>>(temps));
+
+  // Pre-compute two expressions made out of lambda_max and lambda_min, for
+  // speeding up evaluating HLL flux
```

Comment:
Did you check that it is faster? I agree with precomputing these two quantities, but I would remove the second half of this sentence unless if you measured that this is faster.

Reply:
Yes, I agree, and will remove the second line.
By the way, can there be a merit of allocating two precomputed TempScalar if we are not sure that will improve overall performance? (or should it be investigated by actually measuring the performance?)

Reply:
We have measured performance a bit in the past and the single biggest impact on performance is memory allocations. Combining the allocations into a single `Variables` makes a big difference, especially in functions that are called once on each DG interface/mortar, and have relatively low arithmetic compute intensity. memory allocations are a system call (and an expensive one at that), and need to happen serially to some extent (fancy memory allocators like tcmalloc and jemalloc do quite well in a threaded environment). This is because there is only one (maybe two) physical memory device in the computer (multiple RAM sticks usually have a single controller, except on some very high-end hardware, hence the "maybe two"), which means that partitioning that and giving it to different threads and processes needs to be managed in a very controlled (and hence to a large degree serial) manner. This means that as the number of threads increases, memory allocations become increasingly expensive